x86_symbolic/tests/Main.hs

```diff
@@ -70,6 +72,17 @@ ingredients = TT.includingOptions [ TTO.Option (Proxy @SaveSMT)
                                   , TTO.Option (Proxy @SaveMacaw)
                                   ] : TT.defaultIngredients
 
+getRegName ::
+  Ctx.Index (MS.MacawCrucibleRegTypes MX.X86_64) t ->
+  String
+getRegName reg =
+  case Ctx.intIndex (Ctx.indexVal reg) (Ctx.size MXS.x86RegAssignment) of
+    Just (Some i) ->
+      let r = MXS.x86RegAssignment Ctx.! i
+          rName = MS.crucGenArchRegName MXS.x86_64MacawSymbolicFns r
+      in show rName
+    Nothing -> error "impossible"
```

Comment:
Consider using `panic` here instead of `error`.

Reply:
I don't know that it makes a big difference here - The main distinction between `panic` and `error` is the message that gets printed out, especially the confirmation that it is a programmer error (and not, say, a user error). But since this code is only part of the test suite, it's only targeted at Macaw developers, and it's clear from that context that it constitutes a programmer error. In any case, it's real easy to change so I'll go ahead and change it!

Reply:
Ah, `Data.Macaw.Panic` isn't exposed, and `Lang.Crucible.Panic` feels misleading... I might just leave this as `error` if that's OK with you.

Reply:
None of the `Panic` modules in the Macaw repo are exposed, and `Lang.Cruicble.Panic` feels misleading... I might just leave this as `error` if that's alright with you.

Reply:
The usual convention is to create a `PanicComponent` instance for each package so that the `panic` messages can trace back to the specific package that threw the panic (see, for instance, [`Data.Macaw.X86.Symbolic.Panic`](https://github.com/GaloisInc/macaw/blob/1add47389a099b6dedfc6a574a71e2ce433bc617/x86_symbolic/src/Data/Macaw/X86/Symbolic/Panic.hs)).

That being said, I'm fine with `error` as well.

Reply:
Is the suggestion that test-suites should have their own `PanicComponent`? Again, it wouldn't be too much effort to add one, but I'm still having a hard time seeing the advantages of adding this over just using `error`. Could you describe the benefits to doing this? I'm happy to implement this (for all the test-suites in the repo) if we can agree it has benefits.

Comparing a few possible approaches:

- We could try to make this function total (e.g., by introducing a `Maybe`). I would argue there's not much point - it's in a test-suite, so doesn't need to be very robust, and it returning `Nothing` would indicate a bug in the test harness (not in an individual test). Thus, I think partiality is both easier and clearer.
- We could have an incomplete pattern match - I think this is basically the same as using `undefined`, and would throw some kind of generic exception?
- We could abort/exit the process - not sure how this would interact with Tasty (e.g., would it be considered a test failure? would other tests run after it fails?)
- We could use `error`, which raises the [`ErrorCall`](https://hackage.haskell.org/package/ghc-internal-9.1001.0/docs/src/GHC.Internal.Exception.html#ErrorCall) exception
- We could define a custom exception type and throw that, like [`Tasty.HUnit` does](https://hackage.haskell.org/package/tasty-hunit-0.10.0.3/docs/src/Test.Tasty.HUnit.Orig.html#assertFailure)
- We could use `panic`, which [itself defines a custom exception type](https://hackage.haskell.org/package/panic-0.4.0.1/docs/Panic.html#t:Panic). It's relatively undocumented, but my impression is that this exception is meant primarily to ask users to file bug reports. This doesn't seem super relevant to me, as users generally don't run the test suite.

All in all, I don't see many advantages nor disadvantages of any of these approaches over one another, and using `error` is really easy :)

Reply:
> Is the suggestion that test-suites should have their own PanicComponent?

Yes. The only real advantage is that you'd get the nice formatting (along with information about where to submit a bug report) that the `panic` library provides. That being said, that's not super important, since we're in a test suite (and most test suite users are developers).

So, again, I'm fine with `error` as well.

Reply:
Ok, sorry if I went overboard there :smile: I just wanted to make sure I wasn't missing something important!